rfcs/0166-nix-formatting.md

```diff
@@ -0,0 +1,1059 @@
+- Attribute set values are *always* expanded. This has the consequence of always forcing nested attribute sets to be multiline (even if they would be single line otherwise because they only contain a single item), which usually is desired.
+  ```nix
+  {
+    foo.bar.baz = "qux";
+    foo' = {
+      bar.baz = "qux";
+    };
+  }
+  ```
+- "statement-like" expressions like "let", "if" and "assert" always use #4 (or #1).
+- If the value is a `with` followed by a function application or list or attribute set, try to use #3.
+  ```nix
+  buildInputs = with pkgs; [
+    some
+    dependencies
+  ];
+  ```
+
+**Alternatives**
+
+One could eliminate style #2 by having #4 always start on the first line. This would even reduce indentation in some cases. However, this may look really weird in other cases, especially when the binder is very long:
+
+```nix
+some.very.long.attr = callFunction
+  arg1
+  arg2
+  arg3;
+```
+
+#### inherit
+
+**Description**
+
+The items are either all on the same line, or all on a new line each (with indentation).
+
+**Examples**
+
+```nix
+inherit foo bar baz;
+inherit
+  foo'
+  bar'
+  baz'
+;
+```
+
+#### inherit from
+
+**Description**
+
+- If the inherit target is single-line, it is placed on the same line as the `inherit`, even if the following items do not fit onto one line.
+- Otherwise, it starts on a new line with indentation, like the others.
+  - In that case, the remaining items are force-expanded too, even if they would have fit onto one line in the first place.
+
+**Examples**
+
+```nix
+inherit (pkgs) ap1 ap2 ap3;
+inherit (pkgs)
+  app1
+  app2
+  # …
+  app42
+;
+inherit
+  (pkgs.callPackage ./foo.nix {
+    arg = "val";
+  })
+  attr1
+  attr2
+;
+```
+
+#### Semicolon placement
+
+**Description**
+
+The semicolon is always placed on the same line as the expression it concludes.
+
+**Examples**
+
+```nix
+{
+  attr1 = bar;
+  attr2 = function call {
+    # stuff
+  };
+  attr3 =
+    function call
+      many
+      arguments;
+  attr4 =
+    let
+      foo = "bar";
+    in
+    some statement;
+  attr5 =
+    if foo then
+      "bar"
+    else
+      "baz";
+  attr6 =
+    let
+      foo = false;
+    in
+    if foo then "bar" else "baz";
+  attr7 = function (
+    if foo then
+      "bar"
+    else
+      "baz"
+  );
+  attr8 =
+    cond1
+    || cond2
+    ||
+      some function call
+      && cond3;
+}
+```
+
+**Rationale and alternatives**
+
+
+There are four considered semicolon styles:
+1. On a new line without indentation.
+  - This clearly marks a separation between attributes, however it is wasteful of space.
+  ```nix
+  attr3 =
+    function call
+      many
+      arguments;
+  attr3 =
+    let
+      foo = "bar";
+    in
+    some statemens;
+  ```
+2. On a new line with one indentation level.
+  - Just as wasteful on space as (1), but a bit less clear about signaling the end of the binding.
```

Comment:
I'd argue on the basis of low-tech ergonomics that this is preferable over (1): Editors will reasonably put you on the same level of indentation as the previous line, so if you want to terminate the sequence, just hit `;` and `Enter`. It's a reasonable trade for slight visual ambiguity, and it's not wrong to end at the same level of indentation as the expression to be assigned.

I'm strongly in favor in of the quirky-looking but highly diffable "Nix tail" at least for `inherit`. But having that for anything that's not a closing brace is probably good.

Reply:
I don't think we need to optimise for typing code. Formatters are there so you can just type whatever you want and it will format it correctly.

For binders/assignments, we deemed putting the semicolon on its own line as too much deviation from the conventional style, it just looks too weird.

However for `inherit`'s we agreed on this style:
```
inherit (foo)
  bar
  baz
;
```

This section should be rewritten to be clearer about this though. I'd also be in favor of indenting the semicolon here btw, this looks better:

```nix
inherit (foo)
  bar
  baz
  ;
```

Reply:
> I don't think we need to optimise for typing code. Formatters are there so you can just type whatever you want and it will format it correctly.

I'm not arguing for *optimizing* for typing, but as a tiebreaker and background guideline that's in line with this RFC's motivation: reducing friction in the contribution and review process.

Consider the case of people contributing tiny changes through the GitHub web UI: Can we make it more likely that what they type won't get stuck in CI? These will usually be beginners with the Nix ecosystem or coding in general. A red CI check is prone to scare them off, or delay a review indefinitely because they may never get to figure out how to fix it. More advanced contributors may install Git and a plain text editor, but how likely is it that they will have full development environment that happens to do all the right things, including running the formatter?

You may not see these barriers, but they are very real for first-time contributors that may never come again because of some particular papercut that got them off track.

Reply:
Implemented in https://github.com/serokell/nixfmt/pull/118/commits/1446d98049e50bf3cf5589adc0d16ce6de602d8c. Though this change is probably best viewed in the file view rather than as diff.